### PR TITLE
[WIP] gh-112535: Implement fallback implementation of _Py_ThreadId()

### PR DESCRIPTION
**Pull request by corona10**
* Issue: gh-112535

### CODE REVIEW

File: Include/object.h [hunk 1]
```diff
@@ -283,6 +293,13 @@ _Py_ThreadId(void)
     // Both GCC and Clang have supported __builtin_thread_pointer
     // for s390 from long time ago.
     tid = (uintptr_t)__builtin_thread_pointer();
+#elif defined(HAVE_THREAD_ID_FALLBACK)
```
Comment (corona10) [inline Include/object.h]: @colesbury Please let me know, it would be proper explaination.

**Comment by corona10** [comment]
FYI, I checked that it works properly by removing platform-specific implementation in my local.

File: Include/object.h [hunk 2]
```diff
@@ -283,6 +283,19 @@ _Py_ThreadId(void)
...
+#elif defined(thread_local) || defined(__GNUC__)
+    #if defined(thread_local)
+    static thread_local int __tp = 0;
+    #else
+    // Assume that we only support C11 compilers.
+    static __thread int __tp = 0;
+    #endif
+    // gh-112535: Using characteristics of TLS address mapping.
+    // The address of the thread-local variable is not equal with the actual thread pointer,
+    // However, it has the property of being determined by loader at runtime, with no duplication of values
```
Comment (corona10) [inline Include/object.h:298]: I am not sure it's proper explaination. Some elements are initialzied at the startup and some elements are blahblahblah..

File: Include/object.h [hunk 3]
```diff
@@ -283,6 +283,19 @@ _Py_ThreadId(void)
     // Both GCC and Clang have supported __builtin_thread_pointer
     // for s390 from long time ago.
     tid = (uintptr_t)__builtin_thread_pointer();
+#elif defined(thread_local) || defined(__GNUC__)
+    #if defined(thread_local)
+    static thread_local int __tp = 0;
```
Comment (colesbury) [inline Include/object.h]: I don't think this will do what we want.`__tp` will be a unique variable in each C file that includes this header.

**Review by colesbury — COMMENTED** [review]
We want the thread-local variable to be in a C file:

- `_PyThread_Id()` should fall back to a function call that returns the thread id
- You can put that function in `pystate.c` and return the address of `_Py_tss_tstate`. (We can't return the *value* of `_Py_tss_tstate` because that might be NULL)

File: Include/object.h (hunk 3 above)
Comment (colesbury) [reply Include/object.h]: To expand on this: if you declare a `static` variable inside a `static inline` function, then that variable is different in each translation unit (C file) that includes the `static inline` function.

**Pull request closed without merging**